MrkdwnJsFileType.js

```diff
@@ -0,0 +1,252 @@
+
+var MrkdwnJsFileType = function(project) {
+    this.type = "mrkdwn";
+    this.datatype = "mrkdwn";
+    this.project = project;
+    this.API = project.getAPI();
+
+    this.logger = this.API.getLogger("loctool.lib.MrkdwnJsFileType");
+
+    this.extensions = [ ".js" ];
+
+    this.extracted = this.API.newTranslationSet(project.getSourceLocale());
+    this.newres = this.API.newTranslationSet(project.getSourceLocale());
+    this.pseudo = this.API.newTranslationSet(project.getSourceLocale());
+
+    this.pseudos = {};
+
+    // generate all the pseudo bundles we'll need
+    project.settings && project.settings.locales && project.settings.locales.forEach(function(locale) {
+        var pseudo = this.API.getPseudoBundle(locale, this, project);
+        if (pseudo) {
+            this.pseudos[locale] = pseudo;
+        }
+    }.bind(this));
+
+    // for use with missing strings
+    if (!project.settings.nopseudo) {
+        this.missingPseudo = this.API.getPseudoBundle(project.pseudoLocale, this, project);
+    }
+};
+
+var defaultMappings = {
+    "**/*.js": {
+        template: "[dir]/[basename]_[locale].js"
+    }
+};
+
+/**
+ * Return the mapping corresponding to this path.
+ * @param {String} pathName the path to check
+ * @returns {Object} the mapping object corresponding to the
+ * path or undefined if none of the mappings match
+ */
+MrkdwnJsFileType.prototype.getMapping = function(pathName) {
+    if (typeof(pathName) === "undefined") {
+        return undefined;
+    }
+    var mappings, match, mdSettings = this.project.settings && this.project.settings.mrkdwn;
+    if (mdSettings) {
+        mappings = mdSettings.mappings || defaultMappings;
+        var patterns = Object.keys(mappings);
+
+        if (patterns) {
+            match = patterns.find(function(pattern) {
+                return mm.isMatch(pathName, pattern);
+            });
+        }
+    }
+
+    return match && mappings[match];
+}
+
+MrkdwnJsFileType.prototype.getDefaultMapping = function() {
+    return defaultMappings["**/*.js"];
+}
+
+var alreadyLoc = new RegExp(/(^|\/)(([a-z][a-z])(-[A-Z][a-z][a-z][a-z])?(-([A-Z][A-Z])(-[A-Z]+)?)?)\//);
+
+/**
+ * Return true if the given path is an Mrkdwn template file and is handled
+ * by the current file type.
+ *
+ * @param {String} pathName path to the file being questions
+ * @returns {boolean} true if the path is a java file, or false
+ * otherwise
+ */
+MrkdwnJsFileType.prototype.handles = function(pathName) {
+    this.logger.debug("MrkdwnJsFileType handles " + pathName + "?");
+    var extension = path.extname(pathName).toLowerCase();
+    var ret = (this.extensions.indexOf(extension) > -1);
+    var normalized = pathName;
+
+    if (ret) {
+        var match = alreadyLoc.exec(pathName);
+        if (match && match.length > 2) {
+            if (this.API.utils.iso639[match[3]]) {
+                if (match.length < 6 || !match[6] || !this.API.utils.iso3166[match[6]]) {
```

Comment:
The regex.exec() function returns an array of matches, one per parenthesized subexpression. Since the regex has many parentheses, if the match array is greater than two, then the whole regex matched. The 3rd parenthesis has the regex match of language tag of the locale spec, and 6 is the region tag.

Reply:
Thx! Would you mind storing them in variables, so it is self explanatory for the future?
i.e.
```
var isWholeRegexMatch = match && match.length > 2;
var languageMatch = match[3];
var regionMatch = match[6];
```